src/ffi/expression.rs

```diff
@@ -0,0 +1,399 @@
+                        visit(r, predicates);
+                    }
+                    LogicalExpression::Or(l, r) => {
+                        visit(l, predicates);
+                        visit(r, predicates);
+                    }
+                    LogicalExpression::Not(r) => {
+                        visit(r, predicates);
+                    }
+                },
+                Expression::Predicate(p) => {
+                    predicates.push(p);
+                }
+            }
+        }
+
+        visit(self, &mut predicates);
+
+        predicates
+    }
+}
+
+bitflags! {
+    #[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+    #[repr(C)]
+    pub struct BinaryOperatorFlags: u64 /* We can only have no more than 64 BinaryOperators */ {
+        const EQUALS = 1 << 0;
+        const NOT_EQUALS = 1 << 1;
+        const REGEX = 1 << 2;
+        const PREFIX = 1 << 3;
+        const POSTFIX = 1 << 4;
+        const GREATER = 1 << 5;
+        const GREATER_OR_EQUAL = 1 << 6;
+        const LESS = 1 << 7;
+        const LESS_OR_EQUAL = 1 << 8;
+        const IN = 1 << 9;
+        const NOT_IN = 1 << 10;
+        const CONTAINS = 1 << 11;
+
+        const UNUSED = !(Self::EQUALS.bits()
+            | Self::NOT_EQUALS.bits()
+            | Self::REGEX.bits()
+            | Self::PREFIX.bits()
+            | Self::POSTFIX.bits()
+            | Self::GREATER.bits()
+            | Self::GREATER_OR_EQUAL.bits()
+            | Self::LESS.bits()
+            | Self::LESS_OR_EQUAL.bits()
+            | Self::IN.bits()
+            | Self::NOT_IN.bits()
+            | Self::CONTAINS.bits());
+    }
+}
+
+impl From<&BinaryOperator> for BinaryOperatorFlags {
+    fn from(op: &BinaryOperator) -> Self {
+        match op {
+            BinaryOperator::Equals => Self::EQUALS,
+            BinaryOperator::NotEquals => Self::NOT_EQUALS,
+            BinaryOperator::Regex => Self::REGEX,
+            BinaryOperator::Prefix => Self::PREFIX,
+            BinaryOperator::Postfix => Self::POSTFIX,
+            BinaryOperator::Greater => Self::GREATER,
+            BinaryOperator::GreaterOrEqual => Self::GREATER_OR_EQUAL,
+            BinaryOperator::Less => Self::LESS,
+            BinaryOperator::LessOrEqual => Self::LESS_OR_EQUAL,
+            BinaryOperator::In => Self::IN,
+            BinaryOperator::NotIn => Self::NOT_IN,
+            BinaryOperator::Contains => Self::CONTAINS,
+        }
+    }
+}
+
+pub const ATC_ROUTER_EXPRESSION_VALIDATE_OK: i64 = 0;
+pub const ATC_ROUTER_EXPRESSION_VALIDATE_FAILED: i64 = 1;
+pub const ATC_ROUTER_EXPRESSION_VALIDATE_BUF_TOO_SMALL: i64 = 2;
+
+/// Validates an ATC expression against a schema.
+///
+/// # Arguments
+///
+/// - `atc`: a C-style string representing the ATC expression.
+/// - `schema`: a valid pointer to a [`Schema`] object, as returned by [`schema_new`].
+/// - `fields_buf`: a buffer for storing the fields used in the expression.
+/// - `fields_len`: a pointer to the length of `fields_buf`.
```

Comment:
```suggestion
/// - `fields_buf_len`: a pointer to the length of `fields_buf`.
```

I was a bit confused by the name, if this refers to the size of `fields_buf` we should make `fields_buf` part of the name as well.

Reply:
Done.